### ISSUE TITLE
Add touch-model to models vocabulary

### ISSUE THREAD

**Issue by hyphz** [slavapestov/factor#1267]
It would be very helpful to have a touch-model function added to the model class. This would have all the notification effects of set-model, but would not actually input a new value to set.

The reason to use touch-model would be that a part of a complex model was updated in place via reference. If a model's value is a complex datatype containing other complex datatypes (for example, a tuple of sequences), then those lower level complex datatypes can be extracted by reference and changed without needing to call set-model and thus without notifying the model's listeners. The caller could call set-model anyway to trigger notifications, but this requires it to provide the new value for the entire model which means it must reconstruct the entire data structure that was inside the model when it may have only needed a reference to a smaller part of it. touch-model would allow this kind of notification to be given without needed to reconstruct the model value.

**Comment by mrjbq7** [comment]
Can you not call ``notify-connections`` on a model?

**Comment by hyphz** [comment]
Yes, but this doesn't call the update-model hook nor manage locking correctly.

**Comment by hyphz** [comment]
Update-model is the hook a model subclass can override to respond to being changed - set-model deals with calling notify-connections. I have a sort-of bi-arrow already, <partmodel> where I posted the source code to the mailing list, but it has to duplicate code from set-model which is why I asked for a touch-model to be available as a single location for that code. There is still the problem that extracted referenced structures can be updated when the model is locked, but I'm not sure anything can be done about that.

**Comment by mrjbq7** [comment]
I'm not sure what ``touch-model`` does exactly, one way would be to do everything ``set-model`` does except for the ``notify-connections``:

```
: touch-model ( value model -- )
    dup locked?>> [ 2drop ] [
        [ swap >>value update-model ] with-locked-model
    ] if ;
```

Is that what you mean @hyphz ?

**Comment by hyphz** [comment]
No, the reverse. Touch-model would do everything set-model does except for inputting a new value for the model.

**Comment by mrjbq7** [comment]
Oh, you mean ``touch-model`` just triggers the notifications?

Something like this?

```factor
: touch-model ( model -- )
    dup locked?>> [ drop ] [
        [ notify-connections ] with-locked-model
    ] if ;
```

**Comment by hyphz** [comment]
It would need to perform update-model as well, so that hooks on the model ran, but yes. As I said, the idea is for when the data value held within the model is something like a list, where an individual element of the list could be updated in place, thus technically updating the model but not requiring the model's value property to change.

**Comment by mrjbq7** [comment]
Is there a problem with calling ``>>value`` on the model again?  (Aside from perhaps some minor performance and elegance)?

```factor
model [ value>> ] [ set-model ] bi
```
